---
title: "Processing data and stats"
author: "Vincent Manzanilla, INTERTRYP"
output: github_document
---

# Day 2

| Time      | Activity                      | Slides                                | Hands-on                                                |
|-----------|-------------------------------|---------------------------------------|---------------------------------------------------------|
| Morning   | Read Base analysis            |                                       |       [Link here](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/Day2_READBASED_stats.md)                         | 




# Metagenomic read based profiles

In this tutorial we will explore and analyse the results of kraken2 read profiling data using the re-estimatd abundances of bracken. These exercises are meant to show how to conceptually approach your data analysis but there are many more and different ways to explore your data. The most important thing to keep in mind is that you have to understand your own data and analyses. One way to achieve this is to perform visual explorations that help you to judge whether the data are appropriate for your question.

Now let’s start the fun!

In R studio.

Load libraries:

**Data Manipulation and Wrangling**

```{r lib1}
# dplyr: A grammar of data manipulation, providing a consistent set of verbs to solve common data manipulation challenges
library(dplyr)

# tibble: Provides a modern reimagining of data frames, making them more user-friendly
library(tibble)

# tidyverse: A collection of R packages designed for data science, all sharing an underlying design philosophy, grammar, and data structures
library(tidyverse)
```

**Microbiome Analysis**

```{r lib2}
# microbiome: Provides tools for microbiome data analysis and visualization
library(microbiome)

# microbiomeutilities: Extends functionalities of the microbiome package with additional utilities
library(microbiomeutilities)

# vegan: Functions for ecological analysis, including ordination methods, diversity analysis and other functions for community ecologists
library(vegan)

# compositions: Provides methods for analyzing compositional data, which is common in fields such as microbiome research
library(compositions)

# mia: Microbiome Analysis, provides functionalities for microbiome data analysis and visualization
library(mia)

# miaViz: Visualization utilities for microbiome analysis, part of the mia package ecosystem
library(miaViz)

# phyloseq: Microbiome Analysis, provides functionalities for microbiome data analysis and visualization
library(phyloseq)
```

**Data Visualization**

```{r lib3}
# ggplot2: A system for creating elegant and versatile data visualizations based on the grammar of graphics
library(ggplot2)

# hrbrthemes: Contains additional themes, theme components, and utilities for ggplot2
library(hrbrthemes)

# pheatmap: Pretty heatmaps, provides more control over the heatmap visualization
library(pheatmap)

# RColorBrewer: Provides color palettes for visualizing data, particularly useful in ggplot2 visualizations
library(RColorBrewer)

# ggrepel: Extends ggplot2 by adding better text label placement to avoid overlaps
library(ggrepel)

# patchwork: Makes it easier to combine multiple ggplot2 plots into one overall plot layout
library(patchwork)
```

## 1. load the data

First we need to load our data. Usually the biggest bottleneck between raw data and analyses is to get the data in the right shape for your purpose. Often this requires a little bit of data mingling. On this road - google is your best friend to master the R universe :)

Let’s first load the relative abundance table of the bracken results.

For this part we are using the [phyloseq](https://joey711.github.io/phyloseq/) package. The phyloseq package is a tool to import, store, analyze, and graphically display complex phylogenetic sequencing data that has already been clustered into Operational Taxonomic Units (OTUs), especially when there is associated sample data, phylogenetic tree, and/or taxonomic assignment of the OTUs. 

First st your working directory: 

```{r wd}
setwd("~/Documents/project/Metagenomics_2024/Metagenomics_2024/Day_2/")
```

And load your two files:

```{r data}
merged_metagenomes <- phyloseq::import_biom("../DATA//merge_species.biom") 
meta <- read.csv(file = "../DATA/tryp_metadata.csv", sep = ",")
```

#### phyloseq-ize Data

Any data already in an R session can be annoated/coerced to be recognized by phyloseq’s functions and methods. This is important, because there are lots of ways you might receive data related to a microbiome project, and not all of these will come from a popular server or workflow that is already supported by a phyloseq import function. 

![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/phyloseq.png)

- **otu_table** - Works on any numeric matrix. You must also specify if the species are rows or columns
- **tax_table** - Works on any character matrix. The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
- **sample_data** - Works on any data.frame. The rownames must match the sample names in the otu_table if you plan to combine them as a phyloseq-object

 Let's look at the phlyseq object. 

```{r dat1}
merged_metagenomes
```

Summarize the data.  

```{r dat2}
# Summarize the phyloseq object 'merged_metagenomes'
#microbiome::summarize_phyloseq(merged_metagenomes)

# Display the first few rows of the OTU (Operational Taxonomic Unit) table
head(otu_table(merged_metagenomes))

# Display the first few rows of the taxonomy table
head(tax_table(merged_metagenomes))

# Display the first few rows of the sample data associated with 'merged_metagenomes'
#head(sample_data(merged_metagenomes))

# Get the sample variables of the phyloseq object 'merged_metagenomes'
#sample_variables(merged_metagenomes)
```



We need to add the metadata to the phyloseq object. 

## 2. format the data

```{r dat3}
# Sort the 'meta' data frame by the 'SRA.identifier' column
meta <- meta %>% arrange(row_number(SRA.identifier))

# Associate the sorted metadata to the phyloseq object as sample data
merged_metagenomes@sam_data <- sample_data(meta)

# Extract the sample names from the 'meta' data frame
column_name <- meta %>% pull(Sample)

# Associate the extracted sample names to the phyloseq object
sample_names(merged_metagenomes) <- column_name

# Remove the unnecessary 'k_' prefix in the taxonomy data
merged_metagenomes@tax_table@.Data <- substring(merged_metagenomes@tax_table@.Data, 4)

# Rename the columns of the taxonomy table to represent taxonomic ranks
colnames(merged_metagenomes@tax_table@.Data) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
```

## 3. Basic stats

Before we start anything, let’s just check out or data a little bit (sanity check). Never go blind into your analyses.

```{r dat4}
head(psmelt(merged_metagenomes))
```

</details>  

**Q: How many species and samples are detected in our dataset?**

<details>
<summary>
HINT
</summary>

> ntaxa(merged_metagenomes), nsamples.

</details>  

**Q: Look at the taxonomy  is ther some problem with the taxonomy?**

<details>
<summary>
HINT
</summary>

>  We have contaminant in the data. Which one? (head(psmelt(merged_metagenomes)))

</details>  

**Q: what are the taxa with the more abundant reads?**

<details>
<summary>
HINT
</summary>

>  head(otu_table(merged_metagenomes))

</details> 

Extraction of the sample's names. 

```{r dat5}
sample_names(merged_metagenomes)
```

### Aggregation

Microbial species can be called at multiple taxonomic resolutions. We can easily agglomerate the data based on taxonomic ranks. Here, we agglomerate the data at Family level.

```{r dat16}
# Aggregate rare taxa at the family level for the phyloseq object 'merged_metagenomes'
merged_metagenomes_family <- aggregate_rare(merged_metagenomes, level = "Genus", detection = 0/100, prevalence = 0/100)

# Display the dimensionality of the abundances of 'merged_metagenomes_family'
dim(abundances(merged_metagenomes_family))
```

**Q: How many sample and tax do we have now?**

<details>
<summary>
HINT
</summary>

> There are XX taxa and 30 samples, meaning that there are 29 different Phylum level taxonomic groups. Looking at the rowData after agglomeration shows all Enterococcaceae are combined together, and all lower rank information is lost.

```{r dat7}
head(tax_table(merged_metagenomes_family))
```

</details> 


## 4. QC and Pre-process data

Now that we know a little bit about our data we can start the pre-processing. 


###  Library size / read count

Let us check for distribution of number of sequences retained from the Kraken/Bracken approach.

```{r plot, eval=FALSE}
plot_read_distribution(merged_metagenomes, "Type", plot.type = "density")
```
![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/density_reads.png)

*You can try to plot with diffrent metadata.*


Plotting the read count per sample

```{r plot2, eval=FALSE}
df <- psmelt(merged_metagenomes)  %>%  group_by(Sample, Time) %>%  
  summarise(sum_reads = sum(Abundance)) %>% arrange(sum_reads) 

ggplot(df) +
  geom_bar(aes(reorder(Sample, -sum_reads), sum_reads, fill=Time),
           col="red", alpha = .2, stat="identity") 

```


![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/reads_count.png)


### Prevalence - Detection 


Prevalence quantifies the frequency of samples where certain microbes were detected (above a given detection threshold). The prevalence can be given as sample size (N) or percentage (unit interval).

The population prevalence (frequency) at a 1% relative abundance threshold (detection = 1/100 and count = false), can look like this.

```{r pre}
head(prevalence(merged_metagenomes, detection=1/100, sort=TRUE, count=FALSE))
```

The function arguments detection and count can also be used to access, how many samples do pass a threshold for raw counts. Here, the population prevalence (frequency) at the absolute abundance threshold (count=true) at read count 1 (detection = 1) is accessed.

```{r pre1, eval=FALSE}
# Calculate the prevalence of taxa in the phyloseq object 'merged_metagenomes'
# Detection threshold is set to 1, taxa are sorted by prevalence, and count of taxa is returned
head(prevalence(merged_metagenomes, detection = 1, sort = TRUE, count = TRUE))

# Plot the prevalence of taxa at the Phylum level with a detection threshold of 10000
plot_taxa_prevalence(merged_metagenomes, level = "Phylum", detection = 10000)

# Alternative plot: Create a scatter plot of the coefficient of variation (CV) of taxa in 'merged_metagenomes'
# The CV is a measure of relative variability, calculated as the standard deviation divided by the mean
p1 <- plot_taxa_cv(merged_metagenomes, plot.type = "scatter")

# Apply a log10 scale to the x-axis of the scatter plot
p1 + scale_x_log10()
```

Each point corresponds to a different or unique taxon. The y-axis represents the fraction of samples, these taxa are present. The low prevalence suggests there is a low overlap across samples. 

![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/prevalence.png)

**Q: Which taxa is present in all samples and with a high abundance?**

<details>
<summary>
HINT
</summary>

> Homo sapiens

</details> 


#### Contaminant sequences
Samples might be contaminated with exogenous sequences. We have observed 1 contaminants Homo sapiens.


```{r psmelt}
#check 
head(psmelt(subset_taxa(merged_metagenomes, Genus == "Trypanosoma")))
head(psmelt(subset_taxa(merged_metagenomes, Genus == "Homo")))

#Keep only the kingdom of interest
merged_metagenomes <- subset_taxa(merged_metagenomes, Kingdom %in% c("Archaea", "Bacteria", "Fungi", "Viruses"))
```

Now recheck that your data are clean before continuing the analysis. 

**Q: Which taxa is present in all samples and with a high abundance?**

<details>
<summary>
HINT
</summary>

> head(tax_table(merged_metagenomes)) or plot_taxa_prevalence(merged_metagenomes, level="Phylum", detection = 10000)

</details> 

### Run rarefaction curves

Before normalization by sub-sampling, let’s have a look at rarefaction curves, evaluate your sequencing effort and make decisions. 
Rarefaction is a statistical technique employed to evaluate species richness based on sampling results. The process involved subsampling reads without replacement to a defined sequencing depth, thereby creating a standardized library size across samples. Any sample with a total read count less than the defined sequencing depth used to rarefy will be discarded. What we want to see in those curves is that it reaches a plateau meaning no new OTUs are discovered (going up on the Y axis) as sequencing is deeper (the X axis). If this happens we can be pretty confident that we sequenced all the OTUs that were present in the sample.

```{r rare1, eval=FALSE}
install.packages("remotes")
remotes::install_github("gauravsk/ranacapa")
```

```{r rare2, eval=FALSE}
ranacapa::ggrare(pseq, step = 10, color = "Sample", se = FALSE) +
    geom_vline(xintercept = min(sample_sums(pseq)), color = "gray60")
```

![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/rare.png)
 ***Q: What do you observe?***

### Data Normalization

Data transformations are common in (microbial) ecology (Legendre 2001) and used to improve compatibility with assumptions related to specific statistical methods, mitigate biases, enhance the comparability of samples or features, or to obtain more interpretable values.

Examples include the logarithmic transformation, calculation of relative abundances (percentages), and compositionality-aware transformations such as the centered log-ratio transformation (clr).

Let us summarize some commonly used transformations in microbiome data science; further details and benchmarkings available in the references:

- *relabundance* relative transformation; also known as total sum scaling (TSS) and compositional transformation. This converts counts into percentages (at the scale [0, 1]) that sum up to. Much of the currently available taxonomic abundance data from high-throughput assays (16S, metagenomic sequencing) is compositional by nature, even if the data is provided as counts (Gloor et al. 2017).

- *clr* Centered log ratio transformation (Aitchison 1986) is used to reduce data skewness and compositionality bias in relative abundances, while bringing the data to the logarithmic scale. This transformation is frequently applied in microbial ecology (Gloor et al. 2017). However, this transformation only applies to positive values. Usual solution is to add pseudocount, which adds another type of bias in the data. The robust clr transformation (‘rclr’) aims to circumvent the need to add a pseudocount. While the resulting values from these transformations are difficult interpret directly, this transformation may enhance comparability of relative differences between samples. It is part of a broader Aitchison family of transformations; the additive log ratio transformation (`alr’) is also available. The robust clr (“rclr”) is similar to regular clr (see above) but allows data with zeroes and avoids the need to add pseudocount Martino et al. (2019). See library *compositions*. 

- *pa* presence/absence transformation ignores abundances and only indicates whether the given feature is detected above the given threshold (default: 0). This simple transformation is relatively widely used in ecological research. It has shown good performance in microbiome-based classification performance (Giliberti et al. 2022, Karwowska2024).

- *z* Z transformation scales data to zero mean and unit variance; this us used to bring features (or samples) to more comparable levels in terms of mean and scale of the values. This can enhance visualization and interpretation of the data

- *log*, *log2*, *log10* Logarithmic transformations; used e.g. to reduce data skewness; with compositional data the clr (or rclr) transformation is often preferred.

- *hellinger* Hellinger transformation equals to the square root of relative abundances. This ecological transformation can be useful if we are interested in changes in relative abundances.

- *rank* Rank transformation replaces each value by its rank. Also see ‘rrank’ (relative rank transformation). This has use for instance in non-parametric statistics.

- Other available transformations include Chi square (‘chi.square’), Frequency transformation (‘frequency’), and Make margin sum of squares equal to one (‘normalize’)


The data contains read counts. We can convert these into relative abundances and other formats. Compare abundance of a given taxonomic group using the example data before and after the compositionality transformation (with a cross-plot, for instance). You can also compare the results to CLR-transformed data (see e.g. Gloor et al. 2017)

Have a look at the function *?microbiome::transform*.


```{r trans}
microbiome::transform(merged_metagenomes, transform = "compositional")
```

In order to assess the effect of the different transformations, you can use the *clr* or the *log10* for the  downstream analysis. 

## 5. Microbiome composition

Microbial abundances are typically ‘compositional’ (relative) in the current microbiome profiling data sets. This is due to technical aspects of the data generation process (see e.g. Gloor et al., 2017).

The next example calculates relative abundances as these are usually easier to interpret than plain counts. For some statistical models we need to transform the data into other formats as explained in above link (and as we will see later).

*detection*: Detection threshold for absence/presence (percentage reads).
*prevalence*: Prevalence threshold (in [0, 1]) (presence accross samples).

```{r aggre}
pseq <- aggregate_rare(merged_metagenomes, level = "Genus", detection = 0.05/100, prevalence = 20/100)
pseq <- microbiome::transform(pseq, transform = "compositional")
```


**Q: what is the meaning and effect of compositional?**

<details>
<summary>
HINT
</summary>

>  abundances(pseq)

</details>  


#### Visualization

**Bar plot** are useful to have a broad look at the data. 

```{r plot3, eval=FALSE}
# Create a plot of the taxonomic composition at the Family level
p <- plot_composition(pseq,
                      taxonomic.level = "Species", # Specify the taxonomic level for the plot
                      sample.sort = "Sample",     # Sort samples by sample identifier
                      x.label = "Sample",         # Label for the x-axis
                      group_by = "Gut") +          # Group samples by the 'Time' variable

  # Use a color palette from the Color Brewer for filling the Family groups
  scale_fill_brewer("Species", palette = "magma-viridis") +

  # Arrange the legend items in a single column
  guides(fill = guide_legend(ncol = 1)) +

  # Convert the y-axis to represent relative abundance as a percentage
  scale_y_percent() +

  # Add labels and a title to the plot
  labs(x = "Samples", y = "Relative abundance (%)",
       title = "Relative abundance data") + 

  # Apply a clean theme with a horizontal grid
  theme_ipsum(grid = "Y") +

  # Customize the theme for x-axis text and legend text
  theme(axis.text.x = element_text(angle = 90, hjust = 1), # Rotate x-axis text 90 degrees for readability
        legend.text = element_text(face = "italic"),    # Italicize the legend text
        legend.position = "none")

# Print the plot
print(p)
 
```

![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/barplot_family.png)

**Q: Does the profiles look similar between samples? Can you spot any trends?**

<details>
<summary>
HINT
</summary>

> A: We see that the same one or two OTU dominates all samples but there is some variability between samples. We can observe some trend on the Entrococcus proption over time.

</details>  


*If you have time you can also visualize the other taxonomic levels (e.g. species) with the same approach. Try to come up with the code yourself.*

```{r plot4, eval=FALSE}
# Aggregate rare taxa at the Genus level for the phyloseq object 'merged_metagenomes'
# Taxa are included if they have a detection threshold of 0.05% and a prevalence of 20%
pseq <- aggregate_rare(merged_metagenomes, level = "Phylum", detection = 0.05/100, prevalence = 20/100)

# Transform the data to compositional (relative abundance) format
pseq <- microbiome::transform(pseq, transform = "compositional")



n <- dim(tax_table(pseq))[1]
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))



# Create a plot of the taxonomic composition at the Genus level
p <- plot_composition(pseq,
                      taxonomic.level = "Phylum", # Specify the taxonomic level for the plot
                      sample.sort = "Sample",    # Sort samples by sample identifier
                      x.label = "Sample",        # Label for the x-axis
                      otu.sort = "abundance",    # Sort OTUs by their abundance
                      group_by = "Type") +       # Group samples by the 'Type' variable

  # Use a color palette from the Color Brewer for filling the Genus groups
  scale_fill_manual(values = col_vector) +

  # Arrange the legend items in a single column
  guides(fill = guide_legend(ncol = 1)) +

  # Convert the y-axis to represent relative abundance as a percentage
  scale_y_percent() +

  # Add labels and a title to the plot
  labs(x = "Samples", y = "Relative abundance (%)",
       title = "Relative abundance data") + 

  # Apply a clean theme with a horizontal grid
  theme_ipsum(grid = "Y") +

  # Customize the theme for x-axis text and legend text
  theme(axis.text.x = element_text(angle = 90, hjust = 1), # Rotate x-axis text 90 degrees for readability
        legend.text = element_text(face = "italic"))       # Italicize the legend text

# Print the plot
print(p)
```

![](https://github.com/vincentmanz/Metagenomics_2024/blob/main/Day_2/img/barplot_genus.png)

