
# General introduction


Microbial community structure, diversity, and population structure are fundamental aspects to understand evolution, niche adaptation and demographic history of bacterial species. As high-throughput sequencing has become cost effective and accessible, sequencing populations of bacteria across the whole metagenome provides unprecedented resolution to investigate within-host evolution, transmission history and population structure. Moreover, analysis of genetic content of microbial communities through metagenomics has become the mainstream methodology. During this one-week course you will learn and apply bioinformatic techniques to perform population genetics and study microbial communities with metagenomic approaches. The goal is to become familiar with the bioinformatic analysis tools and to be able to utilize them in your own research after the course.

The course with be a mix of lectures and hands-on exercises.
The course will be held in English.
The course unix server will be provided by the course organizers.

# Schedule
The course will be held in PUCE in .... 


## Day 1

* Introduction
* Working with command line
* Read based analysis

## Day 2
2024-06-04 9-16

* Read based analysis with MEGAN
* Read based data analysis in R
* Assembly

## Day 3
2024-06-05 9-16

* Assembly QC
* Genome-resolved metagenomics - Anvio

# Target group
This course is targeted at MSc, PhD, PosDoc and researcher who are interested in using metagenomics for investigating microbial communities.

# Learning outcomes
QA/QC of the raw sequence data, cleaning and assembly; Gene-by-gene annotation; Visualization of data; Taxonomic classification; Functional assignment, Analyses of microbial community composition; Comparative metagenomics; MAG (MetagenomeAssembledGenome) analysis; Statistical methods for metagenome data analysis.

# Requirements
Basic knowledge of Linux/Unix command line and R programming.